File: account-service/src/main/java/com/athena/lms/account/service/TransferService.java
package com.athena.lms.account.service;

import com.athena.lms.account.dto.request.TransferRequest;
import com.athena.lms.account.dto.response.TransferResponse;
import com.athena.lms.account.entity.Account;
import com.athena.lms.account.entity.AccountBalance;
import com.athena.lms.account.entity.AccountTransaction;
import com.athena.lms.account.entity.FundTransfer;
import com.athena.lms.account.event.AccountEventPublisher;
import com.athena.lms.account.repository.AccountBalanceRepository;
import com.athena.lms.account.repository.AccountRepository;
import com.athena.lms.account.repository.AccountTransactionRepository;
import com.athena.lms.account.repository.FundTransferRepository;
import com.athena.lms.common.dto.PageResponse;
import com.athena.lms.common.exception.BusinessException;
import com.athena.lms.common.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.Pageable;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestTemplate;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.Map;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class TransferService {

    private final FundTransferRepository transferRepository;
    private final AccountRepository accountRepository;
    private final AccountBalanceRepository balanceRepository;
    private final AccountTransactionRepository transactionRepository;
    private final AccountEventPublisher eventPublisher;
    private final RestTemplate restTemplate;

    @Value("${lms.product-service.url:http://lms-product-service:8087}")
    private String productServiceUrl;

    @Value("${lms.internal.service-key:}")
    private String serviceKey;

    @Transactional
    public TransferResponse initiateTransfer(TransferRequest req, String tenantId, String initiatedBy) {
        // Idempotency check
        if (req.getIdempotencyKey() != null) {
            var existing = transferRepository.findByReference(req.getIdempotencyKey());
            if (existing.isPresent()) {
                FundTransfer t = existing.get();
                return TransferResponse.from(t,
                        accountRepository.findById(t.getSourceAccountId()).map(Account::getAccountNumber).orElse(null),
                        accountRepository.findById(t.getDestinationAccountId()).map(Account::getAccountNumber).orElse(null));
            }
        }

        // Parse transfer type
        FundTransfer.TransferType transferType;
        try {
            transferType = FundTransfer.TransferType.valueOf(req.getTransferType().toUpperCase());
        } catch (IllegalArgumentException e) {
            throw BusinessException.badRequest("Invalid transfer type: " + req.getTransferType());
        }

        // Resolve source account
        Account sourceAccount = accountRepository.findByIdAndTenantId(req.getSourceAccountId(), tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Source account", req.getSourceAccountId()));
        if (sourceAccount.getStatus() != Account.AccountStatus.ACTIVE) {
            throw new BusinessException(HttpStatus.UNPROCESSABLE_ENTITY, "Source account is " + sourceAccount.getStatus());
        }

        // Resolve destination account
        Account destAccount;
        if (req.getDestinationAccountId() != null) {
            destAccount = accountRepository.findById(req.getDestinationAccountId())
                    .orElseThrow(() -> new ResourceNotFoundException("Destination account", req.getDestinationAccountId()));
        } else if (req.getDestinationAccountNumber() != null) {
            destAccount = accountRepository.findByAccountNumber(req.getDestinationAccountNumber())
                    .orElseThrow(() -> BusinessException.badRequest("Destination account not found: " + req.getDestinationAccountNumber()));
        } else {
            throw BusinessException.badRequest("Either destinationAccountId or destinationAccountNumber is required");
        }

        if (destAccount.getStatus() != Account.AccountStatus.ACTIVE) {
            throw new BusinessException(HttpStatus.UNPROCESSABLE_ENTITY, "Destination account is " + destAccount.getStatus());
        }

        // Same account check
        if (sourceAccount.getId().equals(destAccount.getId())) {
            throw BusinessException.badRequest("Cannot transfer to the same account");
        }

        // Currency check
        if (!sourceAccount.getCurrency().equals(destAccount.getCurrency())) {
            throw BusinessException.badRequest("Currency mismatch: " + sourceAccount.getCurrency() + " vs " + destAccount.getCurrency());
        }

        // For INTERNAL transfers, verify same customer
        if (transferType == FundTransfer.TransferType.INTERNAL
                && !sourceAccount.getCustomerId().equals(destAccount.getCustomerId())) {
            throw BusinessException.badRequest("INTERNAL transfers require same customer; use THIRD_PARTY for different customers");
        }

        // Calculate charge (fail-open: 0 if product-service unreachable)
        BigDecimal chargeAmount = calculateCharge(transferType.name(), req.getAmount(), tenantId);

        BigDecimal totalDebit = req.getAmount().add(chargeAmount);

        // Lock balances in UUID order to prevent deadlocks
        UUID first = sourceAccount.getId().compareTo(destAccount.getId()) < 0
                ? sourceAccount.getId() : destAccount.getId();
        UUID second = first.equals(sourceAccount.getId()) ? destAccount.getId() : sourceAccount.getId();

        AccountBalance firstBal = balanceRepository.findByAccountIdForUpdate(first)
                .orElseThrow(() -> new ResourceNotFoundException("Balance", first));
        AccountBalance secondBal = balanceRepository.findByAccountIdForUpdate(second)
                .orElseThrow(() -> new ResourceNotFoundException("Balance", second));

        AccountBalance sourceBal = first.equals(sourceAccount.getId()) ? firstBal : secondBal;
        AccountBalance destBal = first.equals(destAccount.getId()) ? firstBal : secondBal;

        // Sufficient funds check
        if (sourceBal.getAvailableBalance().compareTo(totalDebit) < 0) {
            throw new BusinessException(HttpStatus.UNPROCESSABLE_ENTITY,
                    "Insufficient funds. Required: " + totalDebit + " " + sourceAccount.getCurrency()
                    + " (transfer: " + req.getAmount() + " + charge: " + chargeAmount + ")");
        }

        // Generate reference
        String reference = req.getIdempotencyKey() != null
                ? req.getIdempotencyKey()
                : "TXF-" + UUID.randomUUID().toString().substring(0, 12).toUpperCase();

        // Debit source
        sourceBal.setAvailableBalance(sourceBal.getAvailableBalance().subtract(totalDebit));
        sourceBal.setCurrentBalance(sourceBal.getCurrentBalance().subtract(totalDebit));
        sourceBal.setLedgerBalance(sourceBal.getLedgerBalance().subtract(totalDebit));
        balanceRepository.save(sourceBal);

        // Credit destination
        destBal.setAvailableBalance(destBal.getAvailableBalance().add(req.getAmount()));
        destBal.setCurrentBalance(destBal.getCurrentBalance().add(req.getAmount()));
        destBal.setLedgerBalance(destBal.getLedgerBalance().add(req.getAmount()));
        balanceRepository.save(destBal);

        // Create transaction records
        AccountTransaction debitTxn = AccountTransaction.builder()
                .tenantId(tenantId)
                .accountId(sourceAccount.getId())
                .transactionType(AccountTransaction.TransactionType.DEBIT)
                .amount(totalDebit)
                .balanceAfter(sourceBal.getAvailableBalance())
                .reference(reference)
                .description("Transfer to " + destAccount.getAccountNumber()
                        + (req.getNarration() != null ? " — " + req.getNarration() : ""))
                .channel("TRANSFER")
                .build();
        transactionRepository.save(debitTxn);

        AccountTransaction creditTxn = AccountTransaction.builder()
                .tenantId(tenantId)
                .accountId(destAccount.getId())
                .transactionType(AccountTransaction.TransactionType.CREDIT)
                .amount(req.getAmount())
                .balanceAfter(destBal.getAvailableBalance())
                .reference(reference)
                .description("Transfer from " + sourceAccount.getAccountNumber()
                        + (req.getNarration() != null ? " — " + req.getNarration() : ""))
                .channel("TRANSFER")
                .build();
        transactionRepository.save(creditTxn);

        // Save transfer record
        FundTransfer transfer = FundTransfer.builder()
                .tenantId(tenantId)
                .sourceAccountId(sourceAccount.getId())
                .destinationAccountId(destAccount.getId())
                .amount(req.getAmount())
                .currency(sourceAccount.getCurrency())
                .transferType(transferType)
                .status(FundTransfer.TransferStatus.COMPLETED)
                .reference(reference)
                .narration(req.getNarration())
                .chargeAmount(chargeAmount)
                .initiatedBy(initiatedBy)
                .completedAt(LocalDateTime.now())
                .build();
        transfer = transferRepository.save(transfer);

        eventPublisher.publishTransferCompleted(transfer.getId(),
                sourceAccount.getId(), destAccount.getId(), req.getAmount(), tenantId);

        log.info("Transfer {} completed: {} {} from {} to {} (charge: {})",
                reference, req.getAmount(), sourceAccount.getCurrency(),
                sourceAccount.getAccountNumber(), destAccount.getAccountNumber(), chargeAmount);

        return TransferResponse.from(transfer,
                sourceAccount.getAccountNumber(), destAccount.getAccountNumber());
    }

    @Transactional(readOnly = true)
    public TransferResponse getTransfer(UUID id, String tenantId) {
        FundTransfer t = transferRepository.findByIdAndTenantId(id, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Transfer", id));
        String srcNum = accountRepository.findById(t.getSourceAccountId())
                .map(Account::getAccountNumber).orElse(null);
        String destNum = accountRepository.findById(t.getDestinationAccountId())
                .map(Account::getAccountNumber).orElse(null);
        return TransferResponse.from(t, srcNum, destNum);
    }

    @Transactional(readOnly = true)
    public PageResponse<TransferResponse> getTransfersByAccount(UUID accountId, String tenantId, Pageable pageable) {
        return PageResponse.from(transferRepository.findByAccountId(tenantId, accountId, pageable)
                .map(TransferResponse::from));
    }

    private BigDecimal calculateCharge(String transferType, BigDecimal amount, String tenantId) {
        try {
            String chargeType = "TRANSFER_" + transferType;
            String url = productServiceUrl + "/api/v1/charges/calculate?transactionType="
                    + chargeType + "&amount=" + amount;

            HttpHeaders headers = new HttpHeaders();
            headers.set("X-Service-Key", serviceKey);
            headers.set("X-Service-Tenant", tenantId);
            headers.set("X-Service-User", "account-service");
            HttpEntity<Void> entity = new HttpEntity<>(headers);

            @SuppressWarnings("unchecked")
            ResponseEntity<Map> resp = restTemplate.exchange(url, HttpMethod.GET, entity, Map.class);
            if (resp.getStatusCode().is2xxSuccessful() && resp.getBody() != null) {
                Object charge = resp.getBody().get("chargeAmount");
                if (charge != null) {
                    return new BigDecimal(charge.toString());
                }
            }
        } catch (Exception e) {
            log.warn("Could not fetch charge from product-service: {}. Using 0 charge.", e.getMessage());
        }
        return BigDecimal.ZERO;
    }
}

File: go-services/internal/account/service/transfer_service.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/shopspring/decimal"
	"go.uber.org/zap"

	"github.com/athena-lms/go-services/internal/account/event"
	"github.com/athena-lms/go-services/internal/account/model"
	"github.com/athena-lms/go-services/internal/account/repository"
	"github.com/athena-lms/go-services/internal/common/dto"
	"github.com/athena-lms/go-services/internal/common/errors"
)

// TransferService provides fund transfer business logic.
// Port of Java TransferService.java.
type TransferService struct {
	repo              *repository.Repository
	publisher         *event.Publisher
	logger            *zap.Logger
	productServiceURL string
	serviceKey        string
}

// NewTransferService creates a new TransferService.
func NewTransferService(repo *repository.Repository, publisher *event.Publisher, logger *zap.Logger,
	productServiceURL, serviceKey string) *TransferService {
	return &TransferService{
		repo:              repo,
		publisher:         publisher,
		logger:            logger,
		productServiceURL: productServiceURL,
		serviceKey:        serviceKey,
	}
}

// TransferRequest is the DTO for initiating a transfer.
type TransferRequest struct {
	SourceAccountID          uuid.UUID       `json:"sourceAccountId"`
	DestinationAccountID     *uuid.UUID      `json:"destinationAccountId,omitempty"`
	DestinationAccountNumber *string         `json:"destinationAccountNumber,omitempty"`
	Amount                   decimal.Decimal `json:"amount"`
	TransferType             string          `json:"transferType"`
	Narration                *string         `json:"narration,omitempty"`
	IdempotencyKey           *string         `json:"idempotencyKey,omitempty"`
}

// TransferResponse is the response DTO for transfers.
type TransferResponse struct {
	ID                       uuid.UUID        `json:"id"`
	SourceAccountID          uuid.UUID        `json:"sourceAccountId"`
	SourceAccountNumber      *string          `json:"sourceAccountNumber,omitempty"`
	DestinationAccountID     uuid.UUID        `json:"destinationAccountId"`
	DestinationAccountNumber *string          `json:"destinationAccountNumber,omitempty"`
	Amount                   decimal.Decimal  `json:"amount"`
	Currency                 string           `json:"currency"`
	TransferType             string           `json:"transferType"`
	Status                   string           `json:"status"`
	Reference                string           `json:"reference"`
	Narration                *string          `json:"narration,omitempty"`
	ChargeAmount             decimal.Decimal  `json:"chargeAmount"`
	InitiatedAt              time.Time        `json:"initiatedAt"`
	CompletedAt              *time.Time       `json:"completedAt,omitempty"`
	FailedReason             *string          `json:"failedReason,omitempty"`
}

func transferResponseFrom(t *model.FundTransfer, srcNum, destNum *string) TransferResponse {
	return TransferResponse{
		ID:                       t.ID,
		SourceAccountID:          t.SourceAccountID,
		SourceAccountNumber:      srcNum,
		DestinationAccountID:     t.DestinationAccountID,
		DestinationAccountNumber: destNum,
		Amount:                   t.Amount,
		Currency:                 t.Currency,
		TransferType:             string(t.TransferType),
		Status:                   string(t.Status),
		Reference:                t.Reference,
		Narration:                t.Narration,
		ChargeAmount:             t.ChargeAmount,
		InitiatedAt:              t.InitiatedAt,
		CompletedAt:              t.CompletedAt,
		FailedReason:             t.FailedReason,
	}
}

// InitiateTransfer executes a fund transfer between two accounts.
func (s *TransferService) InitiateTransfer(ctx context.Context, req TransferRequest, tenantID, initiatedBy string) (*TransferResponse, error) {
	if req.Amount.LessThanOrEqual(decimal.Zero) {
		return nil, errors.BadRequest("amount must be > 0")
	}

	// Idempotency check
	if req.IdempotencyKey != nil {
		existing, err := s.repo.GetTransferByReference(ctx, *req.IdempotencyKey)
		if err == nil {
			srcNum := s.getAccountNumber(ctx, existing.SourceAccountID)
			destNum := s.getAccountNumber(ctx, existing.DestinationAccountID)
			resp := transferResponseFrom(existing, srcNum, destNum)
			return &resp, nil
		}
	}

	// Validate transfer type
	upper := strings.ToUpper(req.TransferType)
	if !model.ValidTransferType(upper) {
		return nil, errors.BadRequest("Invalid transfer type: " + req.TransferType)
	}
	transferType := model.TransferType(upper)

	// Resolve source account
	sourceAccount, err := s.repo.GetAccountByIDAndTenant(ctx, req.SourceAccountID, tenantID)
	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, errors.NotFoundResource("Source account", req.SourceAccountID)
		}
		return nil, err
	}
	if sourceAccount.Status != model.AccountStatusActive {
		return nil, errors.NewBusinessError("Source account is " + string(sourceAccount.Status))
	}

	// Resolve destination account
	var destAccount *model.Account
	if req.DestinationAccountID != nil {
		destAccount, err = s.repo.GetAccountByID(ctx, *req.DestinationAccountID)
		if err != nil {
			if err == pgx.ErrNoRows {
				return nil, errors.NotFoundResource("Destination account", *req.DestinationAccountID)
			}
			return nil, err
		}
	} else if req.DestinationAccountNumber != nil {
		destAccount, err = s.repo.GetAccountByNumber(ctx, *req.DestinationAccountNumber)
		if err != nil {
			if err == pgx.ErrNoRows {
				return nil, errors.BadRequest("Destination account not found: " + *req.DestinationAccountNumber)
			}
			return nil, err
		}
	} else {
		return nil, errors.BadRequest("Either destinationAccountId or destinationAccountNumber is required")
	}

	if destAccount.Status != model.AccountStatusActive {
		return nil, errors.NewBusinessError("Destination account is " + string(destAccount.Status))
	}

	// Same account check
	if sourceAccount.ID == destAccount.ID {
		return nil, errors.BadRequest("Cannot transfer to the same account")
	}

	// Currency check
	if sourceAccount.Currency != destAccount.Currency {
		return nil, errors.BadRequest(fmt.Sprintf("Currency mismatch: %s vs %s", sourceAccount.Currency, destAccount.Currency))
	}

	// For INTERNAL transfers, verify same customer
	if transferType == model.TransferTypeInternal && sourceAccount.CustomerID != destAccount.CustomerID {
		return nil, errors.BadRequest("INTERNAL transfers require same customer; use THIRD_PARTY for different customers")
	}

	// Calculate charge (fail-open: 0 if product-service unreachable)
	chargeAmount := s.calculateCharge(string(transferType), req.Amount, tenantID)
	totalDebit := req.Amount.Add(chargeAmount)

	// Generate reference
	reference := fmt.Sprintf("TXF-%s", strings.ToUpper(uuid.New().String()[:12]))
	if req.IdempotencyKey != nil {
		reference = *req.IdempotencyKey
	}

	// Begin transaction
	tx, err := s.repo.Pool().Begin(ctx)
	if err != nil {
		return nil, err
	}
	defer tx.Rollback(ctx)

	// Lock balances in UUID order to prevent deadlocks
	first, second := sourceAccount.ID, destAccount.ID
	if first.String() > second.String() {
		first, second = second, first
	}

	firstBal, err := s.repo.GetBalanceForUpdate(ctx, tx, first)
	if err != nil {
		return nil, errors.NotFoundResource("Balance", first)
	}
	secondBal, err := s.repo.GetBalanceForUpdate(ctx, tx, second)
	if err != nil {
		return nil, errors.NotFoundResource("Balance", second)
	}

	var sourceBal, destBal *model.AccountBalance
	if first == sourceAccount.ID {
		sourceBal, destBal = firstBal, secondBal
	} else {
		sourceBal, destBal = secondBal, firstBal
	}

	// Sufficient funds check
	if sourceBal.AvailableBalance.LessThan(totalDebit) {
		return nil, errors.NewBusinessError(
			fmt.Sprintf("Insufficient funds. Required: %s %s (transfer: %s + charge: %s)",
				totalDebit.String(), sourceAccount.Currency, req.Amount.String(), chargeAmount.String()))
	}

	// Debit source
	sourceBal.AvailableBalance = sourceBal.AvailableBalance.Sub(totalDebit)
	sourceBal.CurrentBalance = sourceBal.CurrentBalance.Sub(totalDebit)
	sourceBal.LedgerBalance = sourceBal.LedgerBalance.Sub(totalDebit)
	if err := s.repo.UpdateBalance(ctx, tx, sourceBal); err != nil {
		return nil, err
	}

	// Credit destination
	destBal.AvailableBalance = destBal.AvailableBalance.Add(req.Amount)
	destBal.CurrentBalance = destBal.CurrentBalance.Add(req.Amount)
	destBal.LedgerBalance = destBal.LedgerBalance.Add(req.Amount)
	if err := s.repo.UpdateBalance(ctx, tx, destBal); err != nil {
		return nil, err
	}

	// Create transaction records
	debitDesc := "Transfer to " + destAccount.AccountNumber
	if req.Narration != nil {
		debitDesc += " - " + *req.Narration
	}
	transferChannel := "TRANSFER"
	debitTxn := &model.AccountTransaction{
		TenantID:        tenantID,
		AccountID:       sourceAccount.ID,
		TransactionType: model.TransactionTypeDebit,
		Amount:          totalDebit,
		BalanceAfter:    &sourceBal.AvailableBalance,
		Reference:       &reference,
		Description:     &debitDesc,
		Channel:         transferChannel,
	}
	if err := s.repo.CreateTransaction(ctx, tx, debitTxn); err != nil {
		return nil, err
	}

	creditDesc := "Transfer from " + sourceAccount.AccountNumber
	if req.Narration != nil {
		creditDesc += " - " + *req.Narration
	}
	creditTxn := &model.AccountTransaction{
		TenantID:        tenantID,
		AccountID:       destAccount.ID,
		TransactionType: model.TransactionTypeCredit,
		Amount:          req.Amount,
		BalanceAfter:    &destBal.AvailableBalance,
		Reference:       &reference,
		Description:     &creditDesc,
		Channel:         transferChannel,
	}
	if err := s.repo.CreateTransaction(ctx, tx, creditTxn); err != nil {
		return nil, err
	}

	// Save transfer record
	now := time.Now()
	initiator := initiatedBy
	transfer := &model.FundTransfer{
		TenantID:             tenantID,
		SourceAccountID:      sourceAccount.ID,
		DestinationAccountID: destAccount.ID,
		Amount:               req.Amount,
		Currency:             sourceAccount.Currency,
		TransferType:         transferType,
		Status:               model.TransferStatusCompleted,
		Reference:            reference,
		Narration:            req.Narration,
		ChargeAmount:         chargeAmount,
		InitiatedBy:          &initiator,
		CompletedAt:          &now,
	}
	if err := s.repo.CreateTransfer(ctx, tx, transfer); err != nil {
		return nil, err
	}

	if err := tx.Commit(ctx); err != nil {
		return nil, err
	}

	s.publisher.PublishTransferCompleted(ctx, transfer.ID, sourceAccount.ID, destAccount.ID, req.Amount, tenantID)

	s.logger.Info("Transfer completed",
		zap.String("reference", reference),
		zap.String("amount", req.Amount.String()),
		zap.String("currency", sourceAccount.Currency),
		zap.String("from", sourceAccount.AccountNumber),
		zap.String("to", destAccount.AccountNumber),
		zap.String("charge", chargeAmount.String()))

	srcNum := sourceAccount.AccountNumber
	destNum := destAccount.AccountNumber
	resp := transferResponseFrom(transfer, &srcNum, &destNum)
	return &resp, nil
}

// GetTransfer fetches a transfer by ID.
func (s *TransferService) GetTransfer(ctx context.Context, id uuid.UUID, tenantID string) (*TransferResponse, error) {
	t, err := s.repo.GetTransferByIDAndTenant(ctx, id, tenantID)
	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, errors.NotFoundResource("Transfer", id)
		}
		return nil, err
	}
	srcNum := s.getAccountNumber(ctx, t.SourceAccountID)
	destNum := s.getAccountNumber(ctx, t.DestinationAccountID)
	resp := transferResponseFrom(t, srcNum, destNum)
	return &resp, nil
}

// GetTransfersByAccount returns paginated transfers for an account.
func (s *TransferService) GetTransfersByAccount(ctx context.Context, accountID uuid.UUID, tenantID string, page, size int) (dto.PageResponse, error) {
	transfers, total, err := s.repo.ListTransfersByAccount(ctx, tenantID, accountID, size, page*size)
	if err != nil {
		return dto.PageResponse{}, err
	}

	var responses []TransferResponse
	for _, t := range transfers {
		responses = append(responses, transferResponseFrom(t, nil, nil))
	}
	return dto.NewPageResponse(responses, page, size, total), nil
}

func (s *TransferService) getAccountNumber(ctx context.Context, accountID uuid.UUID) *string {
	a, err := s.repo.GetAccountByID(ctx, accountID)
	if err != nil {
		return nil
	}
	return &a.AccountNumber
}

func (s *TransferService) calculateCharge(transferType string, amount decimal.Decimal, tenantID string) decimal.Decimal {
	if s.productServiceURL == "" {
		return decimal.Zero
	}

	chargeType := "TRANSFER_" + transferType
	url := fmt.Sprintf("%s/api/v1/charges/calculate?transactionType=%s&amount=%s",
		s.productServiceURL, chargeType, amount.String())

	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return decimal.Zero
	}
	req.Header.Set("X-Service-Key", s.serviceKey)
	req.Header.Set("X-Service-Tenant", tenantID)
	req.Header.Set("X-Service-User", "account-service")

	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		s.logger.Warn("Could not fetch charge from product-service, using 0 charge", zap.Error(err))
		return decimal.Zero
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		var body map[string]any
		if err := json.NewDecoder(resp.Body).Decode(&body); err == nil {
			if charge, ok := body["chargeAmount"]; ok && charge != nil {
				d, err := decimal.NewFromString(fmt.Sprintf("%v", charge))
				if err == nil {
					return d
				}
			}
		}
	}
	return decimal.Zero
}
